components/script/dom/performance.rs

```diff
@@ -41,6 +103,79 @@ impl Performance {
                            window,
                            PerformanceBinding::Wrap)
     }
+
+    /// Add a PerformanceObserver to the list of observers with a set of
+    /// observed entry types.
+    pub fn add_observer(&self,
+                        observer: &DOMPerformanceObserver,
+                        entry_types: Vec<DOMString>) {
+        self.observers.borrow_mut().push(PerformanceObserver {
+            observer: Root::from_ref(observer),
+            entry_types
+        });
+    }
+
+    /// Remove a PerformanceObserver from the list of observers.
+    pub fn remove_observer(&self, observer: &DOMPerformanceObserver) {
+        let mut observers = self.observers.borrow_mut();
+        let index = match observers.iter().position(|o| &(*o.observer) == observer) {
+            Some(p) => p,
+            None => return,
+        };
+        observers.remove(index);
+    }
+
+    /// Queue a notification for each performance observer interested in
+    /// this type of performance entry and queue a low priority task to
+    /// notify the observers if no other notification task is already queued.
+    ///
+    /// XXX This should be called at some point by the User Timing, Resource
+    ///     Timing, Server Timing and Paint Timing APIs.
+    pub fn queue_entry(&self, entry: &PerformanceEntry) {
+        // Add the performance entry to the list of performance entries that have not
+        // been notified to each performance observer owner, filtering the ones it's
+        // interested in.
+        let _ = self.observers.borrow()
+                              .iter()
+                              .filter(|o| o.entry_types.contains(entry.entry_type()))
+                              .map(|o| o.observer.queue_entry(entry));
+
+        // If no notification task is queued...
+        if !self.pending_notification_observers_task.get() {
+            self.pending_notification_observers_task.set(true);
+            // ... queue a new notification task.
+            let global = self.global();
+            let window = global.as_window();
+            let task_source = window.performance_timeline_task_source();
+            task_source.queue_notification(self, window);
+        }
+    }
+
+    fn notify_observers(&self) {
+        self.pending_notification_observers_task.set(false);
+        let _ = self.observers.borrow().iter().map(|o| o.observer.notify());
```

Comment:
According to https://w3c.github.io/performance-timeline/#queue-a-performanceentry we should be operating on a copy of the list of observers in order to avoid this problem.